File: Frontend/refinery-management-frontend/src/services/routeService.ts
import { API } from './api';

const getRoutes = async () => {
    const response = await API.get('/routes/');
    return response.data;
};

const createRoute = async (routeData: any) => {
    const response = await API.post('/routes/', routeData);
    return response.data;
};

const updateRoute = async (id: number, routeData: any) => {
    const response = await API.patch(`/routes/${id}/`, routeData);
    return response.data;
};

const deleteRoute = async (id: number) => {
    const response = await API.delete(`/routes/${id}/`);
    return response.data;
};

const getRoutesByTruckId = async (truckId: string) => {
    const response = await API.get(`/routes?truck=${truckId}`);
    return response.data;
};

const getRouteByDriver = async (driverEmail: string) => {
    const response = await API.get(`/routes?driver=${driverEmail}`);
    return response.data;
};

export default {
    getRoutes,
    createRoute,
    updateRoute,
    deleteRoute,
    getRoutesByTruckId,
    getRouteByDriver,
};

File: Frontend/refinery_driver_app/lib/services/route_service.dart
// route_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'auth_service.dart';

class RouteService {
  static const String baseUrl = 'http://localhost:8002/api';

  Future<List<dynamic>> getRoutes() async {
    final token = await AuthService().getToken();
    final response = await http.get(
      Uri.parse('$baseUrl/routes/'),
      headers: {'Authorization': 'Bearer $token'},
    );
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Failed to load routes');
    }
  }

  Future<dynamic> createRoute(Map<String, dynamic> routeData) async {
    final token = await AuthService().getToken();
    final response = await http.post(
      Uri.parse('$baseUrl/routes/'),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      },
      body: jsonEncode(routeData),
    );
    if (response.statusCode == 201) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Failed to create route');
    }
  }

  Future<dynamic> updateRoute(int id, Map<String, dynamic> routeData) async {
    final token = await AuthService().getToken();
    final response = await http.patch(
      Uri.parse('$baseUrl/routes/$id/'),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      },
      body: jsonEncode(routeData),
    );
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Failed to update route');
    }
  }

  Future<dynamic> deleteRoute(int id) async {
    final token = await AuthService().getToken();
    final response = await http.delete(
      Uri.parse('$baseUrl/routes/$id/'),
      headers: {'Authorization': 'Bearer $token'},
    );
    if (response.statusCode == 204) {
      return true;
    } else {
      throw Exception('Failed to delete route');
    }
  }

  Future<List<dynamic>> getRoutesByTruckId(String truckId) async {
    final token = await AuthService().getToken();
    final response = await http.get(
      Uri.parse('$baseUrl/routes?truck=$truckId'),
      headers: {'Authorization': 'Bearer $token'},
    );
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Failed to load routes by truck id');
    }
  }

  Future<List<dynamic>> getRouteByDriver(String driverEmail) async {
    final token = await AuthService().getToken();
    final response = await http.get(
      Uri.parse('$baseUrl/routes?driver=$driverEmail'),
      headers: {'Authorization': 'Bearer $token'},
    );
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Failed to load routes by driver');
    }
  }
}
